server/controller/store.go

```diff
@@ -194,101 +193,77 @@ func (s storeCtrl) Delete(stream storev1.StoreService_DeleteServer) error {
+			return status.Errorf(codes.Internal, "failed to receive push referrer request: %v", err)
 		}
 
-		// Validate record and get CID
-		recordCID, err := s.validateAndGetCID(request.GetRecord())
-		if err != nil {
+		// Validate the record reference
+		if err := s.validateRecordRef(request.GetRecordRef()); err != nil {
 			return err
 		}
 
-		pushedRef, err := s.pushRecordToStore(stream.Context(), request.GetRecord())
-		if err != nil {
-			return err
-		}
+		var response *storev1.PushReferrerResponse
+
+		switch request.GetOptions().(type) {
+		case *storev1.PushReferrerRequest_Signature:
+			storeLogger.Debug("Signature referrer request received")
 
-		// Handle signature artifact if provided
-		if request.GetOptions() != nil && request.GetOptions().GetSignature() != nil {
-			if err := s.pushSignatureToStore(stream.Context(), request.GetOptions().GetSignature(), recordCID); err != nil {
-				return err
+			response, err = s.pushSignatureReferrer(stream.Context(), request)
+			if err != nil {
+				return status.Errorf(codes.Internal, "failed to push signature: %v", err)
 			}
-		}
+		default:
+			storeLogger.Debug("Unknown referrer type, skipping")
 
-		// Send the response back via stream
-		response := &storev1.PushWithOptionsResponse{
-			RecordRef: pushedRef,
+			continue
 		}
 
 		if err := stream.Send(response); err != nil {
-			return status.Errorf(codes.Internal, "failed to send push options response: %v", err)
+			return status.Errorf(codes.Internal, "failed to send push referrer response: %v", err)
 		}
 	}
 }
 
-// PullWithOptions retrieves records along with their associated OCI artifacts.
-func (s storeCtrl) PullWithOptions(stream storev1.StoreService_PullWithOptionsServer) error {
-	storeLogger.Debug("Called store controller's PullWithOptions method")
-
-	for {
-		// Receive PullWithOptionsRequest from stream
-		request, err := stream.Recv()
-		if errors.Is(err, io.EOF) {
-			storeLogger.Debug("PullWithOptions stream completed")
-
-			return nil
-		}
+func (s storeCtrl) pushSignatureReferrer(ctx context.Context, request *storev1.PushReferrerRequest) (*storev1.PushReferrerResponse, error) {
+	storeLogger.Debug("Pushing signature referrer", "cid", request.GetRecordRef().GetCid())
 
+	// Try to use signature storage if the store supports it
+	if sigStore, ok := s.store.(interface {
+		PushSignature(context.Context, string, *signv1.Signature) error
+	}); ok {
+		err := sigStore.PushSignature(ctx, request.GetRecordRef().GetCid(), request.GetSignature())
 		if err != nil {
-			return status.Errorf(codes.Internal, "failed to receive pull request: %v", err)
-		}
+			storeLogger.Error("Failed to store signature", "error", err, "cid", request.GetRecordRef().GetCid())
 
-		recordRef := request.GetRecordRef()
-
-		// Validate record reference
-		if err := s.validateRecordRef(recordRef); err != nil {
-			return err
+			return nil, status.Errorf(codes.Internal, "failed to store signature: %v", err)
 		}
 
```

Comment:
same logic here regarding preemptive exit

Reply:
i have noticed that we have extended store interafaces, perhaps we could create common types for these like

```type SignatureStore interface {
	PushSignature()
	VerifySignature()
}
```

then you could check if the store is an instance of this interface, which would make the logic more reusable